### DISCUSSION TITLE
Problem mutating data using SWR and flatlist

### DISCUSSION THREAD

**Discussion by adherb** [nandorojo/solito#273] [Q&A]
I am fetching an array of posts using SWR and then displaying the data using React Native's flatlist.

When a user click's a like button on an item a function is called and the data is mutated on the server and the cache is updated to reflect the updated data.

The issue is when the data is mutated and revalidated it causes the flat list to re-render and the images to flash.

I have tried using useCallback to wrap the function and React.memo to wrap the renderItem, but neither are fixing the issue.

I've also made the keyExtractor a random number, to make sure each item is receiving a unique value.

I have no issue when I use this on the web with map, but for some reason I can't get round this issue using flatlist.

Here is my code:

```
export function PostList() {
  const { mutate } = useSWRConfig();
  
  const { data, error } = useSWR(
    `${process.env.SERVER_API}/posts?userId=1`,
    fetcher
  );

  if (error)
    return (
      <View>
        <Text>failed to load</Text>
      </View>
    );
  if (!data)
    return (
      <View>
        <Text>loading...</Text>
      </View>
    );


  async function likePost(postId) {
    try {
      const response = await fetch(
        `${process.env.SERVER_API}/likePosts`,
        {
          method: "POST",
          body: JSON.stringify({ postId, userId: 1 }),
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      const data = await response.json();
      if (!response.ok) {
        console.log("Something went wrong!");
      }
      console.log("data", data);
    } catch (e) {
      console.log("There was a problem!");
    }
    return data;
  }

  const onClick = useCallback((postId) => {
      const options = {
        rollbackOnError: true,
      };

      mutate(
        `${process.env.SERVER_API}/posts?userId=2`,
        likePost(postId),
        options
      );
  }, []);

  const Item = React.memo(
    ({ id, title, image, categories}) => (
      <>
        <RecipeCard
          id={id}
          title={title}
          image={image}
          onClick={onClick}
        />
      </>
    )
  );

  const renderItem = ({ item, index }: any) => (
    <Item
      index={index}
      id={item.recipe_id}
      title={item.title}
      image={item.image
      categories={item.categories}
    />
  );

  const keyExtractor = () => Math.random();

  return (
    <>
      <FlatList
        data={data}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
      />
    </>
  );
}
```

**Comment by nandorojo** [comment]
> I've also made the keyExtractor a random number, to make sure each item is receiving a unique value.

Never do this! This means that on every render, every item has a new key and the whole list resets. Keys should be consistent across renders. A unique key must be deterministic. It’s the same as map.

This isn’t a solito-specific issue, but a react one. 

Key extractor **must** return a unique ID based on the item. It’s up to you to provide this from the data. If you don’t, the list is told it should flicker.

**Reply by adherb** [reply]
Thanks for the reply Fernando.

I realised this after I made the post and changed it to ```item.post_id``` instead, but unfortunately the issue still persists as when the SWR mutates the key it causes the list to re-render.

Completely get that it's a React issue, but have you ran into and solved this issue before? I can't seem to get it to work using SWR or React Query.

The only fix I currently have is to map the data instead of using a Flatlist.
